### DISCUSSION TITLE
How to make windows open in the correct workspace?

### DISCUSSION THREAD

**Discussion by medwatt** [i3/i3#4872] [Q&A]
I have a small script that opens an application in a given workspace. Here's what the code does:

- User provides a letter `choice` to select the app to open. The user also provides the `workspace` the app should open in.
- The required workspace is first open with `i3-msg "workspace $workspace_name"`
- The app is then opened in the workspace with `bash -c "$cmd_run"`

The issue I am experiencing is that even though I am taken to the desired workspace, the app lunches in the workspace I just left, not in the desired workspace. Why is this happening and how can I fix it?

```bash
choice=$1
workspace_name=$2

# associative array {{{

## array_name=("command name", "command name with arguments")
mail=("geary", "geary")
nvim=("nvim" "kitty nvim")

declare -A array=(
    [m]=mail
    [n]=nvim
)

declare -n list="${array[$choice]}"
cmd_name="${list[0]}"
cmd_run="${list[1]}"

# }}}

# open workspace function {{{
open_workspace () {

    # check if workspace is already open
    is_workspace_open=$(i3-msg -t get_workspaces | jq -r '.[] | select(.name=='\"$workspace_name\"')' | wc -l)

    # if workspace is not open
    if [[ $is_workspace_open = 0 ]]; then
        i3-msg "workspace $workspace_name" # go to workspace
        bash -c "$cmd_run"

    else
        # get the name of the focused workspace
        focused_workspace=$(i3-msg -t get_workspaces | jq -r '.[] | select(.focused==true).name')

        # check if we're not currently in the focused workspace
        if [[ "$focused_workspace" != "$workspace_name" ]]; then
            i3-msg "workspace $workspace_name" # go to workspace

        # check if the window is closed
        elif [[ $(xdotool search --onlyvisible --class "$cmd_name"| wc -l) = 0 ]]; then
            bash -c "$cmd_run"
        fi
    fi
}
# }}}

open_workspace
```

**Comment by Airblader** [comment]
First off, your script is unnecessarily complex: you can ditch the if/else by just always using

```
i3-msg workspace --no-auto-back-and-forth $workspace_name
```

The issue you're seeing is probably a race condition: `i3-msg` exits as soon as the command is issued to the IPC, but that does not guarantee that the workspace has been switched to already by the time you execute the command. You can avoid that by going through i3 entirely:

```
i3-msg workspace --no-back-and-forth $workspace_name, exec $cmd_run
```

Do note that the app opening in that workspace still relies on its implementation of the startup notifications specification unless you stay on the workspace until it is open.

**Reply by medwatt** [reply]
How is my script complex? I'm trying to do the following:

1. If workspace doesn't exist: open workspace and lunch app
2. If workspace exists and we're not in that workspace:
   - if app is not open: go to the workspace and lunch app
   - if app is open: go to the workspace

Running the code you suggested has two issues:

1. It lunches another instance of the app every time.
2. It creates a workspace called `--no-back-and-forth $workspace_name$`. Maybe my version of i3 doesn't have that option.

**Reply by Airblader** [reply]
> Maybe my version of i3 doesn't have that option.

No, I just misspelled the option; I fixed that now.

> It l[a]unches another instance of the app every time.

You still need the logic to check whether the app exists or not, but all the logic around whether or not to switch to the workspace is unnecessary.

**Reply by medwatt** [reply]
> > Maybe my version of i3 doesn't have that option.
>
> No, I just misspelled the option; I fixed that now.
>
> > It l[a]unches another instance of the app every time.
>
> You still need the logic to check whether the app exists or not, but all the logic around whether or not to switch to the workspace is unnecessary.

Thanks for fixing it. I'd be curious to see how you can simplify the code.
